contracts/rbn-staking/GaugeController.vy

```diff
@@ -0,0 +1,635 @@
+
+    ## Remove old and schedule new slope changes
+    # Remove slope changes for old slopes
+    # Schedule recording of initial slope for next_time
+    old_weight_bias: uint256 = self._get_weight(_gauge_addr)
+    old_weight_slope: uint256 = self.points_weight[_gauge_addr][next_time].slope
+    old_sum_bias: uint256 = self._get_sum(gauge_type)
+    old_sum_slope: uint256 = self.points_sum[gauge_type][next_time].slope
+
+    self.points_weight[_gauge_addr][next_time].bias = max(old_weight_bias + new_bias, old_bias) - old_bias
+    self.points_sum[gauge_type][next_time].bias = max(old_sum_bias + new_bias, old_bias) - old_bias
+    if old_slope.end > next_time:
+        self.points_weight[_gauge_addr][next_time].slope = max(old_weight_slope + new_slope.slope, old_slope.slope) - old_slope.slope
+        self.points_sum[gauge_type][next_time].slope = max(old_sum_slope + new_slope.slope, old_slope.slope) - old_slope.slope
+    else:
+        self.points_weight[_gauge_addr][next_time].slope += new_slope.slope
+        self.points_sum[gauge_type][next_time].slope += new_slope.slope
+    if old_slope.end > block.timestamp:
+        # Cancel old slope changes if they still didn't happen
+        self.changes_weight[_gauge_addr][old_slope.end] -= old_slope.slope
+        self.changes_sum[gauge_type][old_slope.end] -= old_slope.slope
+    # Add slope changes for new slopes
+    self.changes_weight[_gauge_addr][new_slope.end] += new_slope.slope
+    self.changes_sum[gauge_type][new_slope.end] += new_slope.slope
+
+    self._get_total()
+
+    self.vote_user_slopes[msg.sender][_gauge_addr] = new_slope
+
+    # Record last action time
+    self.last_user_vote[msg.sender][_gauge_addr] = block.timestamp
+
+    log VoteForGauge(block.timestamp, msg.sender, _gauge_addr, _user_weight)
+
+@external
+def disperse_funds(_total_funding: uint256, _duration: uint256 = WEEK):
+    """
+    @notice Fund all vault deposit staking contracts based on weights
+    @param _total_funding Total Funds to distribute
+    @param _duration Duration over which to distribute the funds
+    """
+    assert msg.sender == self.admin  # dev: admin only
+    assert _total_funding > 0, "Funding must be non-zero!"
+
+    rbn: address = self.token
+
+    rbn_balance: uint256 = ERC20(rbn).balanceOf(self)
+    if _total_funding > rbn_balance:
+      ERC20(rbn).transferFrom(msg.sender, self, _total_funding - rbn_balance)
+
+    # For all the gauges, fund each one based on respective weight
+    for gauge in self.gauges:
+      # weight * total_funding / 1e18.
+      # if weight = 1e18, weight = 1.0
+      gauge_funding_amt: uint256 = self._gauge_relative_weight(gauge, block.timestamp) * _total_funding / MULTIPLIER
+      if(gauge_funding_amt == 0):
+        continue
+      ERC20(rbn).approve(gauge, gauge_funding_amt)
+      VaultStaking(gauge).fund(gauge_funding_amt, _duration)
+      log DisperseFunds(gauge, gauge_funding_amt)
```

Comment:
may be too gas inefficient to log on every gauge, better for us just to look that up from the contract on the UI

Reply:
agree. can always look for [fund event](https://github.com/ribbon-finance/gysr-contracts/blob/c89bb54530c15040ca63a39f394c0ad838b2e6e1/contracts/ERC20BaseRewardModule.sol#L163) which is whenever we fund()